core/nodefactory.lua

```diff
@@ -213,12 +213,24 @@ function nodefactory.nnode:absolute ()
 end
 
 function nodefactory.nnode:outputYourself (typesetter, line)
+  -- See typesetter:computeLineRatio() which implements the currently rather messy
+  -- and probably slightly dubious 'hyphenated' logic.
+  -- Example: consider the word "out-put".
+  -- The node queue therefore contains N(out)D(-)N(put) all pointing to the same
+  -- parent N(output).
   if self.parent and not self.parent.hyphenated then
+    -- When we hit N(out) and are not hyphenated, we output N(output) directly
+    -- and mark it as used, so as to skip D(-)N(put) afterwards.
+    -- I guess this was done to ensure proper kerning (vs. outputting each of
+    -- the nodes separately).
     if not self.parent.used then
       self.parent:outputYourself(typesetter, line)
     end
     self.parent.used = true
   else
+    -- It's possible not to have a parent, e.g. N(word) without hyphenation points.
+    -- Either that, or we have a hyphenated parent but are in the case we are
+    -- outputting one of the elements e.g. N(out)D(-) [line break] N(put).
```

Comment:
> `N(out)D(-) [line break] N(put)` 

Actually this would be `N(out)D(-)` (followed by some G(margin) and other zerohbox'es) on one line, and `D(-)N(put)` (followed by some G(margin) and other zerohbox'es) on the subsequent line. In case it matters to be that precise about it...